src/searching/saddleback_search.rs

```diff
@@ -1,85 +1,255 @@
-// Saddleback search is a technique used to find an element in a sorted 2D matrix in O(m + n) time,
-// where m is the number of rows, and n is the number of columns. It works by starting from the
-// top-right corner of the matrix and moving left or down based on the comparison of the current
-// element with the target element.
 use std::cmp::Ordering;
 
-pub fn saddleback_search(matrix: &[Vec<i32>], element: i32) -> (usize, usize) {
-    // Initialize left and right indices
+/// Custom error type to represent errors related to matrix validation.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MatrixError {
+    NonRectangularInput,
+    NotSorted,
+}
+
+/// Checks if the given matrix (vector of vectors) is sorted row-wise and column-wise.
+///
+/// # Arguments
+///
+/// * `matrix` - A vector of vectors representing the matrix to check.
+///
+/// # Returns
+///
+/// Returns `true` if the matrix is sorted both row-wise and column-wise. Otherwise, returns `false`.
+fn is_sorted(matrix: &[Vec<isize>]) -> bool {
```

Comment:
```suggestion
pub fn is_sorted(matrix: &[Vec<isize>]) -> bool {
```

Imagine that you have some matrix. You know that you will query the position of many elements and you want to check if the matrix is soared only once.

Reply:
I understand this point. By the way, do you know how can we express the tests here? Suppose, we have two functions, one for validate matrix, one for saddleback search like this:
```rs
fn validate_matrix(matrix: &[Vec<isize>]) -> Result<(), MatrixError> {
    // Check if all rows have the same length (rectangular matrix).
    if matrix.iter().any(|row| row.len() != matrix[0].len()) {
        return Err(MatrixError::NonRectangularInput);
    }

    // Check if matrix is sorted.
    if !is_sorted(matrix) {
        return Err(MatrixError::NotSorted);
    }

    Ok(())
}
```
and
```rs
fn saddleback_search(matrix: &[Vec<isize>], element: isize) -> Option<(usize, usize)> {
    if matrix.is_empty() || matrix.iter().all(|row| row.is_empty()) {
        return None;
    }

    let mut left_index = 0;
    let mut right_index = matrix[0].len() - 1;

    while left_index < matrix.len() {
        match element.cmp(&matrix[left_index][right_index]) {
            Ordering::Equal => return Some((left_index, right_index)),
            Ordering::Greater => {
                left_index += 1;
            }
            Ordering::Less => {
                if right_index == 0 {
                    break;
                } else {
                    right_index -= 1;
                }
            }
        }
    }

    None
}
```

Can you propose how we can write tests, @vil02?

Reply:
Can you suggest the detailed changes?